### ISSUE TITLE
Android SocketChannel workaround not working around

### ISSUE THREAD

**Issue by ghost** [eclipse-ee4j/mail#403]
I am using the android-mail 1.6.4 library with activation.

Regarding the section of code Protocol.java:541 (getChannel) and the workaround:

``` java
// XXX - Android is broken and SSL wrapped sockets don't delegate
// the getChannel method to the wrapped Socket
if (socket instanceof SSLSocket) {
   try {
Field f = socket.getClass().getDeclaredField("socket");
f.setAccessible(true);
Socket s = (Socket)f.get(socket);
ret = s.getChannel();
   } catch (Exception ex) {
// ignore anything that might go wrong
   }
}
return ret;
    }
```

socket.getClass().getDeclaredField("socket")
This line returns null and throws the exception:

```
No field socket in class Lcom/android/org/conscrypt/Java8FileDescriptorSocket; (declaration of 'com.android.org.conscrypt.Java8FileDescriptorSocket' appears in /system/framework/conscrypt.jar)
```

I am not sure if this is an issue with a change in android versions, but I am using Android 9 and jdk8 here.

My session and property definitions are:

``` kotlin
    @Provides
    @JvmStatic
    @Singleton
    fun providesMailSession(properties: Properties): Session = Session.getInstance(properties)

    @Provides
    @JvmStatic
    @Singleton
    fun providesMailProperties(): Properties = Properties().apply {
        this["mail.smtp.host"] = "mail.newroztelecom.com"
        this["mail.smtp.port"] = 587
        this["mail.smtp.auth"] = true
        this["mail.smtp.starttls.enable"] = true
        this["mail.imap.host"] = "mail.newroztelecom.com"
        this["mail.imap.port"] = 143
        this["mail.imap.starttls.enable"] = true
        this["mail.imap.usesocketchannels"] = true
    }

    @Provides
    @JvmStatic
    @Singleton
    fun providesIdleManager(es: ExecutorService, properties: Properties) =
        IdleManager(Session.getInstance(properties), es)

    @Provides
    @JvmStatic
    @Singleton
    fun providesMailExecutor(): ExecutorService = Executors.newCachedThreadPool()
```

The IMAPStore is initialized with this session, and connected.

I read that SocketChannels do not support read and write operations so I open the folder with 

``` kotlin
folder.open(javax.mail.Folder.READ_ONLY)
```

And then 
``` kotlin
idleManager.watch(task.folder)
```
throws the exception that the folder is not using SocketChannels  because of the socket being null.

**Comment by bshannon** [comment]
The folder should be opened READ/WRITE.

It sounds like the Android implementation has changed.  I guess I'll need to
track down the source code for the version of Android you're using and see
what's changed.  (I'm not an Android developer so this may take some time
for me to get to.)

**Comment by jmehrens** [comment]
Some random notes around this issue:

Based off of the boolean system property org.conscrypt.useEngineSocketByDefault, it looks like we would have to deal with handling the: https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/ConscryptFileDescriptorSocket.java and https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/ConscryptEngineSocket.java


The super class doesn't implement what we need:
https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/AbstractConscryptSocket.java#L297


There are accessors for getFileDescriptor$ and getChannelId() but not sure if they can be adapted into a SocketChannel.

https://github.com/google/conscrypt/blob/7ec8d2d97b36df7a6f8b59d15062195f0d71bc75/common/src/main/java/org/conscrypt/OpenSSLSocketImpl.java#L77

**Comment by bshannon** [comment]
> Some random notes around this issue:
> 
> Based off of the boolean system property org.conscrypt.useEngineSocketByDefault, it looks like we would have to deal with handling the: https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/ConscryptFileDescriptorSocket.java and https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/ConscryptEngineSocket.java

@jmehrens you're way ahead of me here!  Can you help pull me along?

Can you remind me where to find the current Android Socket source code?

How does this "conscrypt" stuff come into play?

> The super class doesn't implement what we need:
> https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/AbstractConscryptSocket.java#L297
> 
> There are accessors for getFileDescriptor$ and getChannelId() but not sure if they can be adapted into a SocketChannel.

There has to be a SocketChannel around somewhere, since that's how we got the
Socket.  If we can't get from the Socket back to the SocketChannel, that's going to
be a problem.

> https://github.com/google/conscrypt/blob/7ec8d2d97b36df7a6f8b59d15062195f0d71bc75/common/src/main/java/org/conscrypt/OpenSSLSocketImpl.java#L77

This would be so much easier to figure out if I could run Android on my desktop
(easily, not using Android Studio).

Anyway, thanks for the pointers.  I guess it's time to read some code...

**Comment by jmehrens** [comment]
@bshannon  Wait a min,  there is a 'socket' field located here: https://github.com/google/conscrypt/blob/7ec8d2d97b36df7a6f8b59d15062195f0d71bc75/common/src/main/java/org/conscrypt/AbstractConscryptSocket.java#L48

I ran into this before and I think the issue is that you need to modify the IdleManager to walk up the class tree probing for the declared field.  Because the issue with Conscrypt is that they have a deeper class hierarchy. 

for (Class<?> k = socket.getClass(); k != Object.class; k = k.getSuperClass()) {
try {
		Field f = k.getDeclaredField("socket");
		f.setAccessible(true);
		Socket s = (Socket)f.get(socket);
		ret = s.getChannel();
                break;
	    } catch (Exception ex) {
		// ignore anything that might go wrong
	    }
}


===Previous information before finding the socket field=======
Here is the information on conscrypt related to Android:
https://source.android.com/devices/architecture/modular-system/conscrypt

It looks like Conscrypt is using BoringSSL, the Google fork of OpenSSL.  As you can see it is a lot of JNI code: https://github.com/google/conscrypt/blob/master/common/src/main/java/org/conscrypt/NativeCrypto.java

Double check me on this but Android source code appears to be located here:
https://cs.android.com/android/platform/superproject/+/master:libcore/ojluni/src/main/java/java/net/SocketImpl.java;drc=6e00db52c733078764b9316768240863ef384e36;l=49

From what I can tell you can toggle through the branches.

It looks like Conscrypt has tests for socket wrapping deals with a channel:
https://github.com/google/conscrypt/blob/master/testing/src/main/java/org/conscrypt/ChannelType.java

Alternatively, we could maybe ping Conscrypt project and have them implement a SocketChannel in the AbstractConscryptSocket.java.

**Comment by jmehrens** [comment]
@bshannon Another way that we could test would be to use Conscrypt as the security provider for OpenJDK.  We could then just run OpenJDK and use an existing IdleManager test.  That would be a close approximation to what is happening on Android.  I'm looking at how to test with OpenJDK and Android but it will take me some time.

**Comment by bshannon** [comment]
It's a little suspicious that the original user account that created this issue has been deleted...

It would be nice if the person who reported this could test the fix, but lacking that...
@jmehrens how are you coming with your testing, particularly using Android Studio?
I'm just trying to decide whether to include this updated workaround in the 2.0.0
release, or hold it for a 2.0.1 release.

Note that I'd like to finalize the 2.0.0 release in the next few weeks, to show that we're
making progress on Jakarta EE 9.

**Comment by jmehrens** [comment]
@bshannon Yeah I wish the OP could have tested this too :)  I tried building VirtualBox VM with Android Studio and in the end that is not working for testing because of KVM issues.  I'm going to have to start over and build my setup on bare metal.  Doubtful I'll be able to make the timeline.

**Comment by bshannon** [comment]
@jmehrens Have you updated your PR to correspond to your discovery in your updated
comments above?

I'm about ready to take the risk and include the PR in the 2.0.0 release.  If it doesn't work,
we're no worse off than before, and we can apply more fixes in a 2.0.1 release.

**Comment by jmehrens** [comment]
@bshannon The PR is correct and could be merged. I was going to update it with some JUnit tests to show that it works.  That I should be able to get done soon.

I'll look at testing also with Conscrypt on OpenJDK as that shouldn't be as much effort as Android Studio.  Testing on Android is going to take me a lot longer.

**Comment by jmehrens** [comment]
Found an old stackoverflow issue that is probably related to this ticket.

https://stackoverflow.com/questions/47346936/javamail-exception-javax-mail-messagingexception-folder-is-not-using-socketcha/

Took the time to answer that question and shamelessly asked that OP for Android help too.

**Comment by szymonorz** [comment]
Any updates on this? JakartaMail 2.0.0 released a month ago but I don't see an update for Android.

**Comment by jmehrens** [comment]
@szymonorz My understanding is that this will target 2.0.1 release.  The pull request was completed but is waiting on review and merge. https://github.com/eclipse-ee4j/mail/issues/403

**Comment by lukasj** [comment]
hopefully fixed by #441
